### PR TITLE
[Profiler] Restructure inputs and capture TensorLists.

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88653
* #87568
* #87567
* #87566
* #87006
* #86880
* #86854
* #86853
* #86802
* __->__ #87825
* #87133

This PR unifies and rationalizes some of the input representation in Result. The current approach of storing separate types in separate vectors is tedious for two types (Tensors and scalars), but would be even more annoying with the addition of TensorLists. A similar disconnection exists with sizes and strides which the user is also expected to zip with tensor_metadata.

I simplified things by moving inputs to a variant and moving sizes and strides into TensorMetadata. This also forced collection of sizes and strides in python tracer which helps to bring it in line with op profiling. Collection of TensorLists is fairly straightforward; `InputOutputEncoder` already has a spot for them (I actually collected them in the original TorchTidy prototype) so it was just a matter of plumbing things through.

Differential Revision: [D40734451](https://our.internmc.facebook.com/intern/diff/D40734451/)

### CODE REVIEW

File: torch/csrc/profiler/data_flow.cpp [hunk 1]
```diff
@@ -33,62 +68,37 @@ void calculateUniqueTensorIDs(
...
           [&](ExtraFields<EventType::TorchOp>& torch_op) {
-            for (auto& m : torch_op.inputs_.tensor_metadata_) {
-              if (m.has_value()) {
-                insert_tensor(*m);
-              }
+            for (auto& i : torch_op.inputs_) {
+              c10::visit(raw_tensors, i);
             }
           },
-          [&](ExtraFields<EventType::Allocation>& alloc_op) {
```
Comment (slgong-fb) [inline torch/csrc/profiler/data_flow.cpp:55]: we are not catching tensors from allocation calls any more?

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 2]
```diff
@@ -61,11 +61,40 @@ using torch::profiler::impl::ActiveProfilerType;
...
+  for (const auto& i : inputs) {
+    c10::visit(
+        c10::overloaded(
+            [&](const TensorMetadata& t) {
+              shapes.emplace_back(t.sizes_);
+              dtypes.emplace_back(scalarTypeToTypeMeta(t.dtype_).name());
+            },
+            [&](const std::vector<TensorMetadata>&) {
+              shapes.emplace_back();
+              dtypes.emplace_back("TensorList");
```
Comment (slgong-fb) [inline torch/csrc/autograd/profiler_kineto.cpp:84]: n00b question - tensorlist is applied only to operator inputs?

**Review by slgong-fb — APPROVED** [review]
Overall LGTM. compilation error may be coming from python binding?

File: torch/csrc/profiler/data_flow.cpp (hunk 1 above)
Comment (robieta) [reply torch/csrc/profiler/data_flow.cpp:55]: We still are. `[&](auto& i) { raw_tensors(i); }));` forwards to `RawTensors::operator()(ExtraFields<EventType::Allocation>&)`

File: torch/csrc/autograd/profiler_kineto.cpp (hunk 2 above)
Comment (robieta) [reply torch/csrc/autograd/profiler_kineto.cpp:84]: Correct. That's the only place where we have to deal with nested structure.

**Comment by robieta** [comment]
> compilation error may be coming from python binding?

@slgong-fb It was because I moved a ctor into a cpp file, so the class needed `TORCH_API` to link properly.

**Review by chaekit — APPROVED** [review]

**Pull request closed without merging**